### PR TITLE
feat: Support batch receive for consumer.

### PR DESCRIPTION
**Pull request by shibd**
Master Issue: #274

### Motivation

#274  Support batch receive for the consumer.

### Modifications

- Add batch receive policy config on the consumer.
- Add batch receive interface on the consumer.
- Refactor consumer `config set` and `verify logic`. 
  - Since config set logic with verify logic for the consumer is separate, it will be more complicated to read the code and subsequent extensions. 

### Verifying this change
- Add batch receive unit test on Consumer.
- Add more config verify unit test on Consumer.

### Documentation

- [x] `doc-required` 
(Your PR needs to update docs and you will update later)

- [ ] `doc-not-needed` 
(Please explain why)

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: src/ConsumerConfig.cc [hunk 1]
```diff
@@ -265,6 +291,39 @@ ConsumerConfig::ConsumerConfig(const Napi::Object &consumerConfig, pulsar_messag
     }
     pulsar_consumer_configuration_set_dlq_policy(this->cConsumerConfig.get(), &dlq_policy);
   }
+
+  if (consumerConfig.Has(CFG_BATCH_RECEIVE_POLICY) &&
+      consumerConfig.Get(CFG_BATCH_RECEIVE_POLICY).IsObject()) {
+    Napi::Object propObj = consumerConfig.Get(CFG_BATCH_RECEIVE_POLICY).ToObject();
+    int maxNumMessages = -1;
+    if (propObj.Has(CFG_BATCH_RECEIVE_POLICY_MAX_NUM_MESSAGES) &&
```
Comment (gaoran10) [inline src/ConsumerConfig.cc:299]: I found Java client will use the `receiverQueueSize` if `maxNumMessages` is bigger than the `receiverQueueSize`, I'm not sure whether we can add this check in the node client.

File: tests/consumer.test.js [hunk 2]
```diff
@@ -315,6 +346,47 @@ const Pulsar = require('../index');
...
+        });
+
+        const consumer = await client.subscribe({
+          topic: topicName,
+          subscription: 'sub1',
+          subscriptionType: 'Shared',
+          batchReceivePolicy: {
+            maxNumMessages: 10,
+            maxNumBytes: -1,
+            timeoutMs: 500,
```
Comment (gaoran10) [inline tests/consumer.test.js:363]: Could we add a case? Such as sending messages less than 10 to verify timeout works well.

File: src/Consumer.cc [hunk 3]
```diff
@@ -192,36 +193,17 @@ struct ConsumerNewInstanceContext {
...
-    std::string msg("Ack timeout should be 0 or greater than or equal to " +
-                    std::to_string(MIN_ACK_TIMEOUT_MILLIS));
-    deferred->Reject(msg);
-    return deferred->Promise();
-  }
-  int32_t nAckRedeliverTimeoutMs = consumerConfig->GetNAckRedeliverTimeoutMs();
-  if (nAckRedeliverTimeoutMs < 0) {
-    std::string msg("NAck timeout should be greater than or equal to zero");
-    deferred->Reject(msg);
-    return deferred->Promise();
```
Comment (BewareMyPower) [inline src/Consumer.cc:223]: Why did you remove the checks here?

Comment (shibd) [reply src/Consumer.cc:223]: Move to  `consumerConfig->InitConfig` method.

File: src/ConsumerConfig.h [hunk 4]
```diff
@@ -21,14 +21,17 @@
...
 #include "MessageListener.h"
 
 #define MIN_ACK_TIMEOUT_MILLIS 10000
 
 class ConsumerConfig {
  public:
-  ConsumerConfig(const Napi::Object &consumerConfig, pulsar_message_listener messageListener);
+  ConsumerConfig();
   ~ConsumerConfig();
+  void InitConfig(const std::shared_ptr<ThreadSafeDeferred> deferred, const Napi::Object &consumerConfig,
```
Comment (BewareMyPower) [inline src/ConsumerConfig.h:33]: Don't pass const value as the parameter type. For example,

```c++
struct Foo {
  void f1() const {}
  void f2() {}
};

void f(const std::shared_ptr<Foo> foo) {
  foo->f1();
  foo->f2();  // OK, even though f2() is a non-const method
}
```

`f()` can only guarantee `foo` itself cannot be modified like `foo.set(nullptr)`.

However, since this method only needs a reference to the `ThreadSafeDeferred` object and does not transfer or share the ownership of `deferred` to another thread, I suggest passing a reference here.

```c++
  void InitConfig(ThreadSafeDeferred &deferred, const Napi::Object &consumerConfig,
                  pulsar_message_listener messageListener);
```

Then call `deferred.Reject` instead of `deferred->Reject` in the implementation.

BTW, I still don't get why this refactoring is needed in this PR.

File: tests/consumer.test.js (hunk 2 above)
Comment (BewareMyPower) [reply tests/consumer.test.js:363]: +1. You'd better test the `timeoutMs` config works.

File: src/Consumer.cc (hunk 3 above)
Comment (shibd) [reply src/Consumer.cc:223]: Sorry for the late reply. The original code, check settings logic are separated, which will have the following problems:

1. For Some settings, we need to call the C API to determine whether it is successful or not, such as https://github.com/apache/pulsar-client-node/pull/357/files#diff-ca2d7e89de2051db6eacdda40a5abc459724a390b40df60a1e54857caa2ee5f5R321-R325, which we can't do with the original code.
2. After refactoring this way, putting the check settings in a method will make the code clear.

File: src/ConsumerConfig.h (hunk 4 above)
Comment (shibd) [reply src/ConsumerConfig.h:33]: Got it. Thanks for your reminder.

Comment (shibd) [reply src/ConsumerConfig.h:33]: I will keep use std::shared_ptr<ThreadSafeDeferred>, because it will create by NAPI method: `auto deferred = ThreadSafeDeferred::New(info.Env());`

File: src/ConsumerConfig.cc (hunk 1 above)
Comment (shibd) [reply src/ConsumerConfig.cc:299]: This logic has been added on the CPP client:  https://github.com/apache/pulsar-client-cpp/blob/main/lib/ConsumerImplBase.cc#L38-L49

File: tests/consumer.test.js (hunk 2 above)
Comment (shibd) [reply tests/consumer.test.js:363]: I will add it. 

BTW: we shouldn't need to do tests with very complete boundaries on the node client, because it depends on the CPP client, and we've added tests for it in cpp

File: src/ConsumerConfig.cc (hunk 1 above)
Comment (gaoran10) [reply src/ConsumerConfig.cc:299]: Got it, thanks.

**Review by gaoran10 — APPROVED** [review]

**Review by BewareMyPower — APPROVED** [review]
